File: server/feedback/feedback.test.js
const mongoose = require("mongoose");
const request = require("supertest");
const httpStatus = require("http-status");
const chai = require("chai");
const expect = chai.expect;
const app = require("../../index");

chai.config.includeStack = true;

/**
 * root level hooks
 */
after((done) => {
  mongoose.models = {};
  mongoose.modelSchemas = {};
  mongoose.connection.close();
  done();
});

describe("## Feedback APIs", () => {
  let feedback = {
    comment: "Amei a lavagem americana",
    rating: 5,
    service: "624786d8b0cb7760736a3d6f",
    ratingBy: "5c9b8b8f9c9d8b3f8c8b4b9f",
  };

  describe("# POST /api/feedback", () => {
    it("should create a new feedback", (done) => {
      request(app)
        .post("/api/feedback")
        .send(feedback)
        .expect(httpStatus.CREATED)
        .then((res) => {
          expect(res.body.comment).to.equal(feedback.comment);
          expect(res.body.rating).to.equal(feedback.rating);
          expect(res.body.service).to.equal(feedback.service);
          expect(res.body.ratingBy).to.equal(feedback.ratingBy);
          feedback = res.body;
          done();
        })
        .catch(done);
    });
  });

  describe("# GET /api/feedback/:feedbackId", () => {
    it("should get feedback details", (done) => {
      request(app)
        .get(`/api/feedback/${feedback._id}`)
        .expect(httpStatus.OK)
        .then((res) => {
          expect(res.body.comment).to.equal(feedback.comment);
          expect(res.body.rating).to.equal(feedback.rating);
          expect(res.body.service).to.equal(feedback.service);
          expect(res.body.ratingBy).to.equal(feedback.ratingBy);
          done();
        })
        .catch(done);
    });

    it("should report error with message - Not found, when feedback does not exists", (done) => {
      request(app)
        .get("/api/feedback/56c787ccc67fc16ccc1a5e92")
        .expect(httpStatus.NOT_FOUND)
        .then((res) => {
          expect(res.body.message).to.equal("Not Found");
          done();
        })
        .catch(done);
    });
  });

  describe("# Patch /api/feedback/:feedbackId", () => {
    it("should update feedback details", (done) => {
      let update = {
        updates: [
          {
            chave: "comment",
            valor: "Odiei a lavagem americana",
          },
        ],
      };
      request(app)
        .patch(`/api/feedback/${feedback._id}`)
        .send(update)
        .expect(httpStatus.OK)
        .then((res) => {
          expect(res.body.comment).to.equal("Odiei a lavagem americana");
          expect(res.body.rating).to.equal(feedback.rating);
          expect(res.body.service).to.equal(feedback.service);
          expect(res.body.ratingBy).to.equal(feedback.ratingBy);
          done();
        })
        .catch(done);
    });
  });

  describe("# GET /api/feedback/", () => {
    it("should get all feedbacks", (done) => {
      request(app)
        .get("/api/feedback")
        .expect(httpStatus.OK)
        .then((res) => {
          expect(res.body).to.be.an("array");
          done();
        })
        .catch(done);
    });

    it("should get all feedbacks (with limit and skip)", (done) => {
      request(app)
        .get("/api/feedback")
        .query({ limit: 10, skip: 1 })
        .expect(httpStatus.OK)
        .then((res) => {
          expect(res.body).to.be.an("array");
          done();
        })
        .catch(done);
    });
  });

  describe("# GET /api/feedback/service/:serviceId", () => {
    it("should get all feedbacks by service", (done) => {
      request(app)
        .get(`/api/feedback/service/${feedback.service}`)
        .expect(httpStatus.OK)
        .then((res) => {
          expect(res.body).to.be.an("array");
          done();
        })
        .catch(done);
    });
  });
  describe("# GET /api/feedback/rating/:serviceId", () => {
    it("should get average rating by service", (done) => {
      request(app)
        .get(`/api/feedback/rating/${feedback.service}`)
        .expect(httpStatus.OK)
        .then((res) => {
          expect(res.body.media).to.equal(5);
          done();
        })
        .catch(done);
    });
  });

  describe("# GET /api/feedback/user/:userId", () => {
    it("should get all feedbacks by user", (done) => {
      request(app)
        .get(`/api/feedback/user/${feedback.ratingBy}`)
        .expect(httpStatus.OK)
        .then((res) => {
          expect(res.body).to.be.an("array");
          done();
        })
        .catch(done);
    });
  });
  describe("# DELETE /api/feedback/", () => {
    it("should delete feedback", (done) => {
      request(app)
        .delete(`/api/feedback/${feedback._id}`)
        .expect(httpStatus.OK)
        .then((res) => {
          expect(res.body.comment).to.equal("Odiei a lavagem americana");
          expect(res.body.rating).to.equal(feedback.rating);
          expect(res.body.service).to.equal(feedback.service);
          expect(res.body.ratingBy).to.equal(feedback.ratingBy);
          done();
        })
        .catch(done);
    });
    it("should get no content because the feedback already is deleted", (done) => {
      request(app)
        .delete(`/api/feedback/${feedback._id}`)
        .expect(httpStatus.NO_CONTENT)
        .then((res) => {
          done();
        })
        .catch(done);
    });
  });
});

File: testes-integracao/src/test/java/tests/FeedbackTest.java
package tests;

import core.Core;
import org.junit.FixMethodOrder;
import org.junit.Test;
import org.junit.runners.MethodSorters;

import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.*;


@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class FeedbackTest extends Core {
    private static String idFeedback;
    private static String idUser;
    private static String idService;

    @Test
    public void tc01_getAllFeedbacks() {
        given()
                .when()
                .get("/api/feedbacks")
                .then()
                .statusCode(200)
        ;
    }

    @Test
    public void tc02_tryRegisterEmptyFeedback() {

        given()
                .body("{}")
                .when()
                .post("/api/feedbacks")
                .then()
                .statusCode(400)
                .body("message", is(
                        "\"rating\" is required and " +
                        "\"service\" is required and " +
                        "\"ratingBy\" is required"))
        ;
    }

    @Test
    public void tc03_tryRegisterFeedbackSuccessfully() {

        idFeedback = given()
                .body(" {\"comment\": \"Amei a lavagem americana\", " +
                        "\"rating\": 5, " +
                        "\"service\": \"12314123124\"," +
                        "\"ratingBy\": \"123241233\"}")
                .when()
                .post("/api/feedbacks")
                .then()
                .statusCode(201)
                .body("comment", is("Amei a lavagem americana"))
                .body("rating", is(5))
                .body("service", is("12314123124"))
                .body("ratingBy", is("123241233"))
                .extract().path("_id")
        ;
    }

    @Test
    public void tc04_gettingFeedbackByIdSuccessfully() {
        given()
                .when()
                .get("/api/feedbacks/" + idFeedback)
                .then()
                .statusCode(200)
                .body("comment", is("Amei a lavagem americana"))
                .body("rating", is(5))
                .body("service", is("12314123124"))
                .body("ratingBy", is("123241233"))
        ;
    }

    @Test
    public void tc05_changeFeedbackSuccessfully() {
        given()
                .body("{\"updates\": " +
                        "[{\"chave\": \"comment\"," +
                        "\"valor\": \"Odiei a lavagem americana\"}," +
                        "{\"chave\": \"rating\"," +
                        "\"valor\": 1 \n}]}")
                .when()
                .patch("/api/feedbacks/" + idFeedback)
                .then()
                .statusCode(200)
                .body("comment", is("Amei a lavagem americana"))
                .body("rating", is(1))
        ;
    }

    @Test
    public void tc06_deleteFeedbackRegistered() {
        given()
                .when()
                .delete("/api/feedbacks/" + idFeedback)
                .then()
                .statusCode(200)
        ;

        given()
                .when()
                .get("/api/feedbacks" + idFeedback)
                .then()
                .statusCode(404)
                .body("message", is("Not Found"))
        ;

    }

    @Test
    public void tc07_tryDeleteFeedbackNotExist() {
        given()
                .when()
                .delete("/api/feedbacks" + idFeedback)
                .then()
                .statusCode(404)
                .body("message", is("Not Found"));
    }

    @Test
    public void tc07_getFeedbacksByUser() {
        given()
                .when()
                .delete("/api/feedbacks/user/" + idUser)
                .then()
                .statusCode(200)
    }
    
    @Test
    public void tc07_getFeedbacksByService() {
        given()
                .when()
                .delete("/api/feedbacks/service/" + idService)
                .then()
                .statusCode(200)
    }
}
